src/v2/upward.rs

```diff
@@ -139,38 +144,48 @@ impl<A> Upward<A> {
     {
         match self {
             Upward::Known(t) => f(t),
-            Upward::Unknown => default(),
+            Upward::Unknown(_) => default(),
         }
     }
 
-    /// Converts from `&Option<A>` to `Option<&A>`.
-    pub const fn as_ref(&self) -> Upward<&A> {
+    /// Converts from `&Upward<A, R>` to `Upward<&A, &R>`.
+    pub const fn as_ref_with_residual(&self) -> Upward<&A, &R> {
         match *self {
             Self::Known(ref x) => Upward::Known(x),
-            Self::Unknown => Upward::Unknown,
+            Self::Unknown(ref r) => Upward::Unknown(r),
         }
     }
 
     /// Returns [`Upward::Unknown`] if the option is [`Upward::Unknown`],
     /// otherwise calls `f` with the wrapped value and returns the result.
-    pub fn and_then<U, F>(self, f: F) -> Upward<U>
+    pub fn and_then<U, F>(self, f: F) -> Upward<U, R>
     where
-        F: FnOnce(A) -> Upward<U>,
+        F: FnOnce(A) -> Upward<U, R>,
     {
         match self {
-            Upward::Unknown => Upward::Unknown,
+            Upward::Unknown(r) => Upward::Unknown(r),
             Upward::Known(x) => f(x),
         }
     }
 }
 
-impl<A, E> Upward<Result<A, E>> {
+impl<A> Upward<A> {
+    /// Converts from `&Upward<A>` to `Upward<&A>`.
+    pub const fn as_ref(&self) -> Upward<&A> {
+        match *self {
+            Self::Known(ref x) => Upward::Known(x),
+            Self::Unknown(_) => Upward::Unknown(()),
```

Comment:
seems we are loosing the residual here.

```suggestion
            Self::Unknown(ref r) => Upward::Unknown(r),
```